sys/drv/bcm2835_pic.c

```diff
@@ -0,0 +1,204 @@
+  bcm2835_pic_state_t *bcm2835_pic = ie->ie_source;
+  unsigned irq = ie->ie_irq;
+
+  bus_size_t reg = BCM2835_INTC_IRQBDISABLE;
+  if (irq < BCM2835_INT_GPU1BASE)
+    reg = BCM2835_INTC_IRQ1DISABLE;
+  else if (irq < BCM2835_INT_BASICBASE)
+    reg = BCM2835_INTC_IRQ2DISABLE;
+
+  uint32_t en = in4(reg);
+  out4(reg, en & ~(1 << irq));
+}
+
+static void bcm2835_pic_enable_irq(intr_event_t *ie) {
+  bcm2835_pic_state_t *bcm2835_pic = ie->ie_source;
+  unsigned irq = ie->ie_irq;
+
+  bus_size_t reg = BCM2835_INTC_IRQBENABLE;
+  if (irq < BCM2835_INT_GPU1BASE)
+    reg = BCM2835_INTC_IRQ1ENABLE;
+  else if (irq < BCM2835_INT_BASICBASE)
+    reg = BCM2835_INTC_IRQ2ENABLE;
+
+  uint32_t en = in4(reg);
+  out4(reg, en | (1 << irq));
+}
+
+static resource_t *bcm2835_pic_alloc_intr(device_t *pic, device_t *dev, int rid,
+                                          unsigned irq, rman_flags_t flags) {
+  bcm2835_pic_state_t *bcm2835_pic = pic->state;
+  rman_t *rman = &bcm2835_pic->rm;
+  return rman_reserve_resource(rman, RT_IRQ, rid, irq, irq, 1, 0, flags);
+}
+
+static void bcm2835_pic_release_intr(device_t *pic, device_t *dev,
+                                     resource_t *r) {
+  resource_release(r);
+}
+
+static const char *bcm2835_pic_intr_name(unsigned irq) {
+  const char *type = "BASIC";
+  if (irq < BCM2835_INT_GPU1BASE)
+    type = "GPU0";
+  else if (irq < BCM2835_INT_BASICBASE)
+    type = "GPU1";
+  return kasprintf("%s source %u", type, irq);
+}
+
+static void bcm2835_pic_setup_intr(device_t *pic, device_t *dev, resource_t *r,
+                                   ih_filter_t *filter, ih_service_t *service,
+                                   void *arg, const char *name) {
+  bcm2835_pic_state_t *bcm2835_pic = pic->state;
+  unsigned irq = resource_start(r);
+  assert(irq < BCM2835_NIRQ);
+
+  if (bcm2835_pic->intr_event[irq] == NULL)
+    bcm2835_pic->intr_event[irq] =
+      intr_event_create(bcm2835_pic, irq, bcm2835_pic_disable_irq,
+                        bcm2835_pic_enable_irq, bcm2835_pic_intr_name(irq));
+
+  r->r_handler = intr_event_add_handler(bcm2835_pic->intr_event[irq], filter,
+                                        service, arg, name);
+}
+
+static void bcm2835_pic_teardown_intr(device_t *pic, device_t *dev,
+                                      resource_t *irq) {
+  intr_event_remove_handler(irq->r_handler);
+}
+
+static int bcm2835_pic_map_intr(device_t *pic, device_t *dev, phandle_t *intr,
+                                int icells) {
+  if (icells != 2)
+    return -1;
+
+  /* Interrupt tuples are of the form `<intrtype irq_offset>`. */
+  bcm2835_intrtype_t type = intr[0];
+  unsigned irq_off = intr[1];
+
+  if (type >= BCM2835_INTRTYPE_CNT)
+    return -1;
+  if (irq_off >= BCM2835_NIRQPERTYPE)
+    return -1;
+
+  unsigned irq = irq_off;
+  if (type == BCM2835_INTRTYPE_GPU1)
+    irq += BCM2835_NIRQPERTYPE;
+  else if (type == BCM2835_INTRTYPE_BASIC)
+    irq += BCM2835_NIRQPERTYPE * 2;
```

Comment:
ditto

Reply:
I added a comment to explain what happens here and where is the description of the layout.